xpublish/plugins/hooks.py

```diff
@@ -0,0 +1,100 @@
+from typing import Callable, Iterable, List, Optional
+
+import cachey
+import pluggy
+import xarray as xr
+from fastapi import APIRouter
+from pydantic import BaseModel
+
+from ..dependencies import get_cache, get_dataset, get_dataset_ids, get_plugin_manager, get_plugins
+
+hookspec = pluggy.HookspecMarker('xpublish')
+hookimpl = pluggy.HookimplMarker('xpublish')
+
+
+class Dependencies(BaseModel):
+    dataset_ids: Callable[..., List[str]] = get_dataset_ids
+    dataset: Callable[..., xr.Dataset] = get_dataset
+    cache: Callable[..., cachey.Cache] = get_cache
+    plugins: Callable[..., 'Plugin'] = get_plugins
+    plugin_manager: Callable[..., pluggy.PluginManager] = get_plugin_manager
+
+    def __hash__(self):
+        """Dependency functions aren't easy to hash"""
+        return 0
+
+
+class Plugin(BaseModel):
+    """
+    Xpublish plugins provide ways to extend the core of xpublish with
+    new routers and other functionality.
+
+    To create a plugin, subclass ``Plugin` and add attributes that are
+    subclasses of `PluginType` (`Router` for instance).
+
+    The specific attributes correspond to how Xpublish should use
+    the plugin.
+    """
+
+    name: str
+
+    def __hash__(self):
+        """Make sure that the plugin is hashable to load with pluggy"""
+        things_to_hash = []
+
+        for e in self.dict():
+            if isinstance(e, list):
+                things_to_hash.append(tuple(e))
+            else:
+                things_to_hash.append(e)
+
+        return hash(tuple(things_to_hash))
+
+    def __dir__(self) -> Iterable[str]:
+        """We need to override the dir as pluggy will otherwise try to inspect it,
+        and Pydantic has marked it class only
+
+        https://github.com/pydantic/pydantic/pull/1466
+        """
+        d = super().__dir__()
+
+        d.remove('__signature__')
+
+        return d
+
+
+class PluginSpec(Plugin):
+    """Plugin extension points"""
+
+    @hookspec
+    def app_router(self, deps: Dependencies) -> APIRouter:
```

Comment:
Just for my learning, all these methods are optional right? I.e. a plugin could implement one or more of these hooks?

Reply:
Yep, they are all optional. I'm thinking I might explore mix-ins for common plugin patterns, in addition to a whole lot of documentation around plugins and how the whole ecosystem can work.